support-models/src/main/scala/com/gu/support/workers/states/CreateZuoraSubscriptionState.scala

```diff
@@ -1,31 +1,100 @@
 package com.gu.support.workers.states
 
-import java.util.UUID
+import com.gu.i18n.Country
 
-import com.gu.salesforce.Salesforce.SalesforceContactRecords
+import java.util.UUID
+import com.gu.salesforce.Salesforce.{SalesforceContactRecords, SfContactId}
+import com.gu.support.encoding.Codec.deriveCodec
 import com.gu.support.promotions.PromoCode
 import com.gu.support.workers.{PaymentMethod, SalesforceContactRecord, User, _}
 import org.joda.time.LocalDate
-import com.gu.support.encoding.CustomCodecs.{decodeLocalTime, encodeLocalTime}
+import com.gu.support.encoding.CustomCodecs.{decodeLocalTime, encodeLocalTime, decodeCountry, encodeCountryAsAlpha2}
+import com.gu.support.encoding.{Codec, DiscriminatedType}
 import com.gu.support.redemptions.RedemptionData
+import com.gu.support.workers.GiftRecipient.{DigitalSubscriptionGiftRecipient, WeeklyGiftRecipient}
 
-case class CreateZuoraSubscriptionState(
+case class CreateZuoraSubscriptionWrapperState(
+  productSpecificState: CreateZuoraSubscriptionState,
   requestId: UUID,
   user: User,
-  giftRecipient: Option[GiftRecipient],
   product: ProductType,
   analyticsInfo: AnalyticsInfo,
-  paymentMethod: Either[PaymentMethod, RedemptionData],
   firstDeliveryDate: Option[LocalDate],
   promoCode: Option[PromoCode],
-  salesforceContacts: SalesforceContactRecords,
   acquisitionData: Option[AcquisitionData]
 ) extends FailureHandlerState
 
-import com.gu.support.encoding.Codec
-import com.gu.support.encoding.Codec._
+object CreateZuoraSubscriptionWrapperState {
+  implicit val codec = deriveCodec[CreateZuoraSubscriptionWrapperState]
+}
+
+sealed trait CreateZuoraSubscriptionState
 
 object CreateZuoraSubscriptionState {
-  import com.gu.support.encoding.CustomCodecs._
-  implicit val codec: Codec[CreateZuoraSubscriptionState] = deriveCodec
+
+  case class CreateZuoraSubscriptionContributionState(
```

Comment:
How do you feel about dropping `CreateZuoraSubscription` from the start of all these class names? They are all inside the `CreateZuoraSubscriptionState` object so it may not really be needed and it does make the names very verbose.

Reply:
It would be nice if we could genericise these classes in some way like 
`CreateZuoraSubscriptionState[Product, ReaderType, AcquisitionType]` but that's not going to give us the varying structures we want is it?

Reply:
yep I think the verboseness is definitely out of control.  I will try that - as long as we don't end up clashing too much with the preceding state and next state.
Regarding the genericising it,  I agree that there is commonality across many of them and the we could represent those similarities in some way, e.g. more complex hierarchy with `trait DeliverableSubscription` or a deeper tree structure and all the rest, that would increase the reuse and communicate more to the compiler, but having gone down that route on the last refactor it ended up being over complicated.
I quite like the nice flat self contained set of products and I think it does get most of the benefit without being totally confusing for new people.  But have a think on it and there may well be a middle ground.